Update lintapidiff

shindere opened the pull request:
While working on `tools/Makefile`, I came across the `lintapidiff` tool
and discovered it did no longer compile. Investigating a bit further showed
it has actually been broken since OCaml 4.08.

This PR fixes the tool so that it compiles again but does not try to
address the result of `make lintapidiff` which runs it on the codebase.

the update has been done incrementally, version by version. Thus, one
branch per vesion is available on `shindere/ocaml`, from `4.08` to `5.0`,
in case there is the wish to include a working verison of the tool in the
previous versions of OCaml or jsut to use it on them.

At the moment, each version branch contains several commits but I could
squash them into one on each branch if there is the wish to cherry-pick,
to make this easier.

dra27 commented on .gitignore:
```diff
@@ -308,6 +308,8 @@ META
 /tools/make_opcodes.ml
 /tools/ocamltex
 /tools/eventlog_metadata
+/tools/lintapidiff
```
`tools/lintapidiff` is not built (at least at present)

Octachron commented on tools/lintapidiff.ml:
```diff
@@ -126,13 +126,14 @@ module Ast = struct
       | Pmty_signature lst -> List.fold_left self map lst
       | Pmty_functor ({txt;_}, _, m) ->
-          let path = Path.Papply(path, Path.Pident (Ident.create txt)) in
+          let new_name = Ident.create_local txt in
```
The identifier should be `Ident.create_persistent` since we want identifier to share paths between different versions of the standard library.

Octachron commented on tools/lintapidiff.ml:
```diff
@@ -173,7 +176,7 @@ module Ast = struct
       let id =
         orig |> Filename.chop_extension |> Filename.basename |>
-        String.capitalize_ascii |> Ident.create in
+        String.capitalize_ascii |> Ident.create_local in
```
Same as before: this should be `Ident.create_persistent`.

Octachron commented on tools/lintapidiff.ml:
```diff
@@ -125,20 +129,26 @@ module Ast = struct
-      enter_path  path m.pmd_name m.pmd_type m.pmd_attributes map
+      let name =
+        match m.pmd_name.txt with
+        | None -> { m.pmd_name with txt = "NONE"}
```
I think it is better to fail with an `unsupported` here rather than trying to give an unique name to all unamed functor arguments.

Octachron commented:
After some fixes, it seems that the tool might be useful once again which seems better than tracking by hand the presence or absence of `since` annotation. I would propose to update the list of excluded in the main `Makefile` to include the `templates` file to decrease the noise as a start. Currently, the output is still noisy but that could be fixed later.

shindere commented:
Florian Angeletti (2022/10/21 04:46 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -126,13 +126,14 @@ module Ast = struct
>        match ty.pmty_desc with
>        | Pmty_signature lst -> List.fold_left self map lst
>        | Pmty_functor ({txt;_}, _, m) ->
> -          let path = Path.Papply(path, Path.Pident (Ident.create txt)) in
> +          let new_name = Ident.create_local txt in
>
> The identifier should be `Ident.create_persistent` since we want
> identifier to share paths between different versions of the standard
> library.

Thanks, fixed.

shindere commented:
Florian Angeletti (2022/10/21 04:46 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -173,7 +176,7 @@ module Ast = struct
>      try
>        let id =
>          orig |> Filename.chop_extension |> Filename.basename |>
> -        String.capitalize_ascii |> Ident.create in
> +        String.capitalize_ascii |> Ident.create_local in
>
> Same as before: this should be `Ident.create_persistent`.

Also addressed, thanks.

shindere commented:
Florian Angeletti (2022/10/21 04:49 -0700):
> @Octachron commented on this pull request.
>
>
>
> >      in
>      let enter_path path name ty attrs map =
>        let path = Path.Pdot (path, name.txt) in
>        let inherits = f inherits name.loc attrs in
>        add_module_type path ty (inherits, map)
>      in
>      let add_module map m =
> -      enter_path  path m.pmd_name m.pmd_type m.pmd_attributes map
> +      let name =
> +        match m.pmd_name.txt with
> +        | None -> { m.pmd_name with txt = "NONE"}
>
> I think it is better to fail with an `unsupported` here rather than
> trying to give an unique name to all unamed functor arguments.

Done so. At the moment I used `failwith` but another possibility would
be to introduce a dedicated `Unsupported` exception with a string
argument.

shindere commented:
Florian Angeletti (2022/10/21 04:52 -0700):
> After some fixes, it seems that the tool might be useful once again
> which seems better than tracking by hand the presence or absence of
> `since` annotation. I would propose to update the list of excluded in
> the main `Makefile` to include the `templates` file to decrease the
> noise as a start. Currently, the output is still noisy but that could
> be fixed later.

So, I took all the comments into account, rebased on latest trunk and
squashed all the commits. As agreed with @Octachron, I did not backport
the fixes to my version-specific branches because when `lintapidiff` is
run from `trunk` it actually does already check the older branches, so
making it work on older branches did not seem a priority.

For the sake of clarity, I propose not to include the fix of the
***@***.***` annotation in the present PR, so that things remain clearly
separate from each other.

Octachron commented on tools/lintapidiff.ml:
```diff
@@ -147,7 +160,8 @@ module Ast = struct
           ) map lst
     | Psig_exception e ->
-        add_path ~f "exception" path e.pext_name e.pext_attributes inherits map
+        add_path ~f "exceptionu" path e.ptyexn_constructor.pext_name
```
exceptionu -> exception

Octachron commented on tools/lintapidiff.ml:
```diff
@@ -27,6 +27,12 @@
 open Location
 open Parsetree
 
+exception Unsupported of string
```
The `Unsupported` exception is currently unused and should be removed.

Octachron commented on tools/lintapidiff.ml:
```diff
@@ -125,19 +131,26 @@ module Ast = struct
-      | Pmty_functor ({txt;_}, _, m) ->
-          let path = Path.Papply(path, Path.Pident (Ident.create txt)) in
+      | Pmty_functor ((Named ({txt = Some txt0},_)),m) ->
+          let new_name = Ident.create_persistent txt0 in
```
`new_name` is maube questionable as a `name` since the name part of the identifier is the same as before. Maybe more `let ident`?

shindere commented:
Florian Angeletti (2022/10/25 07:47 -0700):
> -        add_path ~f "exception" path e.pext_name e.pext_attributes inherits map
> +        add_path ~f "exceptionu" path e.ptyexn_constructor.pext_name
>
> exceptionu -> exception

Fixed, thanks!

shindere commented:
Florian Angeletti (2022/10/25 07:49 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -27,6 +27,12 @@
>  open Location
>  open Parsetree
>
> +exception Unsupported of string

Sorry, removed.

shindere commented:
Florian Angeletti (2022/10/25 07:51 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -125,19 +131,26 @@ module Ast = struct
>        let self = add_item ~f path inherits in
>        match ty.pmty_desc with
>        | Pmty_signature lst -> List.fold_left self map lst
> -      | Pmty_functor ({txt;_}, _, m) ->
> -          let path = Path.Papply(path, Path.Pident (Ident.create txt)) in
> +      | Pmty_functor ((Named ({txt = Some txt0},_)),m) ->
> +          let new_name = Ident.create_persistent txt0 in
>
> `new_name` is maube questionable as a `name` since the name part of
> the identifier is the same as before. Maybe more `let ident`?

Yes, sure, you are right, thanks a lot for the suggestion, the change
has been done.

Octachron reviewed (approved):
Fixing the compilation of `lintapidiff` seem like a clear improvement to me.
There are few bugs or missing feature that decrease its usability , but the tool is usable and we can fix the remaining issues later.

dra27 reviewed (approved):
LGTM too, apart from reverting the change to `.gitignore`. I double-checked that the warnings produced on the 4.06 build (the last time the tool was used) are equivalently produced in the trunk version (having massaged `VERSIONS` not to include the more recent tags)

shindere committed:
Update lintapidiff

Make the tool compile

shindere commented:
Many thanks for your support and approval, @Octachron!

shindere commented:
David Allsopp (2022/10/26 01:34 -0700):
> @dra27 approved this pull request.
>
> LGTM too, apart from reverting the change to `.gitignore`. I
> double-checked that the warnings produced on the 4.06 build (the last
> time the tool was used) are equivalently produced in the trunk version
> (having massaged `VERSIONS` not to include the more recent tags)

Thanks for your care!

> > @@ -308,6 +308,8 @@ META
>  /tools/make_opcodes.ml
>  /tools/ocamltex
>  /tools/eventlog_metadata
> +/tools/lintapidiff
>
> `tools/lintapidiff` is not built (at least at present)

I fixed .gitignore in 21f9993f1fe64e942011c39a988ce857258c112a which I
had to push separately, sorry about that. I did intent to include it in
the PR but for got to `git add`.
